### PR TITLE
Use WeakConcurrentMap

### PR DESCRIPTION
**Pull request by tylerbenson**
Instead of `Collections.synchronizedMap(new WeakHashMap<>())`.

### CODE REVIEW

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+public class WeakMapManager {
```
Comment (tylerbenson) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java]: @raphw with this PR, we are using quite a few instances of WeakConcurrentMap, so I created this class to allow the cleaning to be done by a single thread, rather than have a thread per map.  Does this look like a reasonable way of managing that?  Any suggestions?

Thanks for providing `WeakConcurrentMap`!

File: dd-java-agent/agent-bootstrap/agent-bootstrap.gradle [hunk 2]
```diff
@@ -9,10 +9,13 @@ excludedClassesConverage += ['datadog.trace.bootstrap.*']
 
 dependencies {
   compile project(':dd-trace-api')
+  compile group: 'com.blogspot.mydailyjava', name: 'weak-lock-free', version: '0.13'
```
Comment (mar-kolya) [inline dd-java-agent/agent-bootstrap/agent-bootstrap.gradle]: Before I dig deeper into this PR I would like to ask a question: why this is better than Guava's `new MapMaker().weakKeys().makeMap()` which seems to do its own clean up and is easier to use.
So just from complexity perspective I think using Guava's implementation is better - unless there are drastic advantages.

Comment (raphw) [reply dd-java-agent/agent-bootstrap/agent-bootstrap.gradle]: Doing "its own cleanup" means that it applies inline expunction. This lib offers the same but it is not recommended if you do not want to interfere with the user thread and add additional latency. Also, Guava is a pretty heavy library and due to its generic functionality, its weak map it adds overhead. This, in brief, is the motivation for this small lib.

Comment (mar-kolya) [reply dd-java-agent/agent-bootstrap/agent-bootstrap.gradle]: @raphw thanks for your reply!

I should say we already have Guava dependency - so just using Guava's version should not introduce any more dependencies.
If I'm reading it correctly Guava's inline expunction has limits on how much work it does in one go so additional latency introduced by it should be bounded.
As far as additional overhead - I guess this is a valid point in general. Although in our particular case this might be not so big concern overall - I mean Guava is probably not terribly slow and we plan to create number of entries that is comparable to number of Span's we have - which overall is not very large.

Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/agent-bootstrap.gradle]: @mar-kolya I think there is some desire to reduce our dependency on guava and potentially remove it altogether.

In general, I think this is more simple and offers more control than guava.

**Comment by tylerbenson** [comment]
Tests aren't passing.  I'm trying to figure out the classpath issues.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    try {
+      Runtime.getRuntime()
+          .addShutdownHook(
+              new Thread() {
+                @Override
+                public void run() {
+                  try {
+                    cleaner.shutdownNow();
+                    cleaner.awaitTermination(5, TimeUnit.SECONDS);
```
Comment (mar-kolya) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java]: at 1 second frequency 5 seconds seems overkill... would you object to just using `CLEAN_FREQUENCY_SECONDS` here?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                }
+              });
+    } catch (final IllegalStateException ex) {
+      // The JVM is already shutting down.
+    }
+  }
+
+  public static <K, V> WeakConcurrentMap<K, V> newWeakMap() {
+    final WeakConcurrentMap<K, V> map = new WeakConcurrentMap<>(false);
+    maps.add(map);
```
Comment (mar-kolya) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java]: This would keep reference to created may forever. This is fine for static things, this might be less fine for dynamic things - and it looks like this is called form places not declared `static`.

At the very least this behavior should be documented. But ideally this should 'do the right thing' - which I guess in this case we be using weakly referenced storage for `maps` as well.

Also, it if turns out that these maps are created/destroyed after initial init stage (i.e. during normal 'request processing') runtime - you may want to rethink use of `CopyOnWriteArrayList` for storage.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java (hunk 3 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java]: 5 seconds is an upper bound.  It should terminate much sooner than that.

File: dd-java-agent/agent-bootstrap/agent-bootstrap.gradle [hunk 5]
```diff
@@ -9,10 +9,13 @@ excludedClassesConverage += ['datadog.trace.bootstrap.*']
...
   compile deps.opentracing
   compile deps.slf4j
   compile group: 'org.slf4j', name: 'slf4j-simple', version: versions.slf4j
   // ^ Generally a bad idea for libraries, but we're shadowing.
+
+  testCompile project(':dd-java-agent:testing')
```
Comment (mar-kolya) [inline dd-java-agent/agent-bootstrap/agent-bootstrap.gradle]: Could you please explain why is this needed?

Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/agent-bootstrap.gradle]: https://github.com/DataDog/dd-trace-java/pull/428/files#diff-05ae743560b75e6d0cea4ff90f545cf4R30

Comment (mar-kolya) [reply dd-java-agent/agent-bootstrap/agent-bootstrap.gradle]: Thanks for explaining. I've tried removing it and build still worked for me for whatever reason - I guess I should have cleaned it out.
Anyway, reason why you need it makes sense, but this dependency is still kind of awkward - maybe it's worth adding a comment here why we need it so we could later review this and maybe pull out some general testing utils so they do not depend on the rest of tooling/bootstrap code?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java (hunk 4 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java]: I reimplemented it using a weak reference to ensure they are GC'd if the returned map is GC'd.

File: dd-java-agent/agent-bootstrap/agent-bootstrap.gradle (hunk 5 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/agent-bootstrap.gradle]: The latest version has a similar problem but with `agent-tooling`.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                    // Don't bother waiting then...
+                  }
+                }
+              });
+    } catch (final IllegalStateException ex) {
+      // The JVM is already shutting down.
+    }
+  }
+
+  public static <K, V> WeakConcurrentMap<K, V> newWeakMap() {
```
[the remark this answers is not in the document]
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMapManager.java]: I've reimplemented it with several options for underlying implementations.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+  }
+
+  static class Guava implements WeakMap.Supplier {
+
+    @Override
+    public <K, V> WeakMap<K, V> get() {
+      return new Adapter<>(new MapMaker().weakKeys().<K, V>makeMap());
+    }
+
+    private static class Adapter<K, V> implements WeakMap<K, V> {
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: You may want to consider making `WeakMap.Supplier.Adapter` public (and potentially renaming it to `WeakMap.MapAdapter` - it seems reasonable that many concurrent maps would try to adhere to common map interface which means that that code would be possible to reuse - at least here.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+class WeakMapSuppliers {
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: Is there a reason for not making this public - other classes in this package are public.

And on similar note - would you consider splitting this file into per-provider files maybe in a separate package (e.g. datadog.trace.agent.tooling.weakmap.suppliers) - just to separate concerns.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  // * `WeakConcurrentMap`: anecdotally used by other agents
+  // * `Guava`: specifically agent use is unknown at the moment, but Guava is a well known library
+  //            backed by big company with many-many users
+
+  /**
+   * Provides instances of {@link WeakConcurrentMap} and retains weak reference to them to allow a
+   * single thread to clean void weak references out for all instances. Cleaning is done every
+   * second.
+   */
+  static class WCMManaged implements WeakMap.Supplier {
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: Again, this being package-private seems odd - could you please explain the reason for this?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+      try {
+        Runtime.getRuntime()
+            .addShutdownHook(
+                new Thread() {
+                  @Override
+                  public void run() {
+                    try {
+                      cleaner.shutdownNow();
+                      cleaner.awaitTermination(5, TimeUnit.SECONDS);
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: Minor nitpick: ideally it would be nice to make this a constant so this is not buried deep in the code.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+  //            backed by big company with many-many users
+
+  /**
+   * Provides instances of {@link WeakConcurrentMap} and retains weak reference to them to allow a
+   * single thread to clean void weak references out for all instances. Cleaning is done every
+   * second.
+   */
+  static class WCMManaged implements WeakMap.Supplier {
+    private static final long CLEAN_FREQUENCY_SECONDS = 1;
+    private static final Queue<WeakReference<WeakConcurrentMap>> maps =
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: Would you object to having all static final constants` names in capitals?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+                      // Don't bother waiting then...
+                    }
+                  }
+                });
+      } catch (final IllegalStateException ex) {
+        // The JVM is already shutting down.
+      }
+    }
+
+    public static void cleanMaps() {
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: I think the only reason this is public is so you could use this in a test. But logic in that test seems slightly flawed because it assumes that cleaner will not kick in in the middle of the test - which it can because we have cleaner thread running.

Instead that test could be rewritten to hold on to `Object` reference util after first `awaitGC` - then null it and wait for some reasonable amount of time for size to eventually become zero - potentially periodically calling `awaitGC`.

Once that is changed you may want to consider making this private and even moving this to `Cleaner.run` to remove one level of indirection.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java (hunk 9 above)
Comment (mar-kolya) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: I think `WCM` is kind of obscure naming and is hard to remember. Would you object to having `WeakConcurrentMapManaged` - this will be typed like 3 times in whole codebase, so typing doesn't increase by much. And FWIW you already have `WeakConcurrentMapSupplierTest`

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+
+    private static class Cleaner implements Runnable {
+
+      @Override
+      public void run() {
+        cleanMaps();
+      }
+    }
+
+    private static class Adapter<K, V> implements WeakMap<K, V> {
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: As far as I can see you repeat this implementation twice - would it make sense to move it 'level up' and use in both places?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/TracerInstaller.java [hunk 14]
```diff
@@ -9,6 +10,8 @@
 public class TracerInstaller {
   /** Register a global tracer if no global tracer is already registered. */
   public static synchronized void installGlobalTracer() {
+    registerWeakMapProvider();
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/TracerInstaller.java]: We use `WeakMap`s quite early on - my guess is that `AgentInstaller` would use them through `ClassLoaderMatcher` - when that class is loaded and static members are initialized. From this perspective changing provider this late may be confusing and fragile (change of the order of loading classes may change implementations used by different parts of the agent).

I guess there are two ways of solving this:
* Move everything into bootstrap - which obviously requires solving problem of repackaging things
* Remove implicit default provider - make it 'blow up' if we try to get a map before provider was explicitly registered - and then fix all things that fail by moving things around (e.g. `JDBCMaps` should probably go closer to instrumentation that actually uses it, and get injected as a helper). This would make sure that all `WeakMap`s are of the same implementation type.

Thoughts?

Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/TracerInstaller.java]: I've moved the registration to earlier on in the process.  That should resolve the issue.  If you want me to make the default throw an error, I can do that too.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java (hunk 8 above)
Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: I don't really consider this part of the public API.  I'd like to encapsulate things better when possible, especially since we're still evaluating which one we want to use.

I thought about putting it in a separate package, but they all feel related to me.  Eventually I expect all but one to be removed.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java (hunk 9 above)
Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: I've updated the name to make it less obscure.

File: dd-trace-ot/src/main/java/datadog/opentracing/DDTracer.java [hunk 15]
```diff
@@ -48,7 +49,7 @@
 
 /** DDTracer makes it easy to send traces and span to DD using the OpenTracing API. */
 @Slf4j
-public class DDTracer implements io.opentracing.Tracer {
+public class DDTracer implements io.opentracing.Tracer, Closeable {
```
[the remark this answers is not in the document]
Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/opentracing/DDTracer.java]: I think this is useful for people that create `DDTracer` instances directly to signal that it should be closed to clean up threads and such.

File: dd-trace-ot/src/main/java/datadog/opentracing/PendingTrace.java [hunk 16]
```diff
@@ -239,6 +239,7 @@ static void close() {
           public Thread newThread(final Runnable r) {
             final Thread thread = new Thread(r, "dd-span-cleaner");
             thread.setDaemon(true);
+            thread.setPriority(Thread.MIN_PRIORITY);
```
[the remark this answers is not in the document]
Comment (tylerbenson) [reply dd-trace-ot/src/main/java/datadog/opentracing/PendingTrace.java]: While I understand the scenario you're referring to, it seems rather benign to me and I expect to rarely happen.  If you're really not comfortable, I can revert this change.  Just seems like a good idea to me.

[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-trace-ot/src/main/java/datadog/opentracing/PendingTrace.java]: Well, I'll leave it up to you. I guess the question you may want to ask yourself: will you be comfortable with this change next time you investigate memory leak in concurrency instrumentation?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/TracerInstaller.java (hunk 14 above)
Comment (mar-kolya) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/TracerInstaller.java]: I personally would prefer to 'fail soon and loud' in this case but I'll defer final decision to you.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java (hunk 12 above)
Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: Do you mind if I leave it as-is for now?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java (hunk 13 above)
Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: done.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java (hunk 12 above)
Comment (mar-kolya) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/WeakMapSuppliers.java]: Would you please consider https://github.com/DataDog/dd-trace-java/pull/436 - it addresses this and also improves (I hope) test coverage.
